src/VisualStudio/CSharp/Impl/LanguageService/CSharpCodeCleanupFixer.cs

```diff
@@ -320,37 +386,92 @@ private async Task<bool> FixTextBufferAsync(TextBufferCodeCleanUpScope textBuffe
+                }
+            }
+        }
 
-                    foreach (var diagnostic in allDiagnostics.Diagnostics)
-                    {
-                        foreach (var diagnosticId in diagnostic.DiagnosticIds)
-                        {
-                            if (context.EnabledFixIds.IsFixIdEnabled(diagnosticId))
-                            {
-                                enabedDiagnosticSets.Add(diagnostic);
-                                break;
-                            }
-                        }
-                    }
+        private async Task<Solution> FixSolutionAsync(Solution solution, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            progressTracker.AddItems(solution.ProjectIds.Count);
+            foreach (var projectId in solution.ProjectIds)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
 
-                    var isRemoveUnusedUsingsEnabled = context.EnabledFixIds.IsFixIdEnabled(RemoveUnusedImportsFixId);
-                    var isSortUsingsEnabled = context.EnabledFixIds.IsFixIdEnabled(SortImportsFixId);
-                    var enabledDiagnostics = new EnabledDiagnosticOptions(enabedDiagnosticSets.ToImmutableArray(),
-                        new OrganizeUsingsSet(isRemoveUnusedUsingsEnabled, isSortUsingsEnabled));
+                var project = solution.GetProject(projectId);
+                progressTracker.Description = project.Name;
 
-                    var newDoc = await codeCleanupService.CleanupAsync(
-                        document, enabledDiagnostics, progressTracker, cancellationToken).ConfigureAwait(true);
+                // FixProjectAsync reports progress for documents within a project, but we limit progress reporting for
+                // a solution to the current project.
+                var projectProgressTracker = new ProgressTracker();
 
-                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-                    return document.Project.Solution.Workspace.TryApplyChanges(newDoc.Project.Solution, progressTracker);
+                var newProject = await FixProjectAsync(project, enabledFixIds, projectProgressTracker, cancellationToken).ConfigureAwait(false);
+                solution = newProject.Solution;
+                progressTracker.ItemCompleted();
+            }
+
+            return solution;
+        }
+
+        private async Task<Project> FixProjectAsync(Project project, FixIdContainer enabledFixIds, ProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            if (project.LanguageServices.GetService<ICodeCleanupService>() == null)
+            {
+                return project;
+            }
+
+            progressTracker.AddItems(project.DocumentIds.Count);
+            foreach (var documentId in project.DocumentIds)
```

Comment:
>foreach (var documentId in project.DocumentIds) [](start = 12, length = 47)

So seems that we fix everything sequentially document by document here. Our current FixAll in project and FixAll in solution performs computes fixes in parallel, at least the batch fixer does, and performs only single fix (instead of multiple being done here), but the operation is still _extremely_ slow for FixAll in solution for mid sized solutions. I suspect this is going to be even slower, and potentially unusable for large size solutions. I feel we should dogfood this feature on real world solutions to verify the performance characteristics before we turn this on.

Reply:
I think document code cleanup does use FixAll to fix everything at once (we improved existing FixAll to support code clean up case), we might need to improve FixAll service again to support this case as well?

I am okay starting from this as we did for document case. but making document code clean up to use fix all rather than doing fix by fix improvement per from like 10s seconds to less than 1 second perf on certain document with all fix enabled.

we might get similar perf improvement by using FixAll service since it can use all existing perf work we did for project/solution wide fixes. for example, going through this way per document will cause us to calcuate diagnostics (if it is not already up to date) per document instead of whole project/solution. and fix won't be done concurrently and etc.

Reply:
anyway, I am fine it being a future improvement unless someone has a strong opinion that it must be done now. right now, I am not sure how many people will actually use so, it would be fine until we get feedback.